File: Array/subarray_with_sum_k.py
def subarray_with_sum_k(arr, k):
    n = len(arr)
    sum = 0
    no_of_subarrays = 0
    prefix_sum_count = {0: 1}
    for i in range(n):
        sum += arr[i]
        if prefix_sum_count.get(sum - k):
            no_of_subarrays += prefix_sum_count.get(sum - k)
        prefix_sum_count[sum] = prefix_sum_count.get(sum, 0) + 1
    return no_of_subarrays


if __name__ == "__main__":
    arr = list(map(int, input(f"Enter array elements separated by space: ").split()))
    k = int(input(f"Enter the value of k: "))
    no_of_subarrays = subarray_with_sum_k(arr, k)
    print(f"Number of subarray whose sum is {k} are {no_of_subarrays}")

File: JavaSolutions/Leetcode/src/Array/SubarrayWithSumK.java
package Array;

import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class SubarrayWithSumK {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter array of integers separated by space:");
        String inputLine = sc.nextLine();
        String[] numbersAsString = inputLine.split(" ");
        int[] numbers = new int[numbersAsString.length];
        for (int i = 0; i < numbersAsString.length; i++) {
            numbers[i] = Integer.parseInt(numbersAsString[i]);
        }
        System.out.println("Enter the value of k: ");
        int k = sc.nextInt();
        System.out.println("Total number of subarray whose sum is " + k + " are: " + findSubArrays(numbers, k));
        sc.close();
    }

    private static int findSubArrays(int[] numbers, int k) {
        int count = 0;
        int prefixSum = 0;
        Map<Integer, Integer> prefixSumCount = new HashMap<>();
        prefixSumCount.put(0, 1);
        for (int num : numbers) {
            prefixSum += num;
            count += prefixSumCount.getOrDefault(prefixSum - k, 0);
            prefixSumCount.put(prefixSum, prefixSumCount.getOrDefault(prefixSum, 0) + 1);
        }
        return count;
    }

}
